scanner/scanner.go

```diff
@@ -202,11 +231,168 @@ func scanNot(s *Scanner) (token.Token, []byte) {
 }
 
 func scanDoubleQuote(s *Scanner) (token.Token, []byte) {
+	t := token.String
+	next, rOffset := scanInDoubleQoutes(s)
+	switch next {
+	case '@', '$', '{':
+		t = token.StringPart
+		s.pushCtx(stateInDoubleQoutes)
+	case '"':
+		s.next()
+	}
+	off := s.offset - rOffset
+	if off > s.begin+1 {
+		off--
+	}
+	return t, s.src[s.begin+1 : off]
+}
+
+func scanInDoubleQoutes(s *Scanner) (byte, int) {
+	nEscape := 0
 	for s.char != '"' && s.err == nil {
+		if s.char == '#' {
+			next := s.peek(2)[1]
+			if next == '@' || next == '$' || next == '{' {
+				return next, nEscape
+			}
+		}
+
+		c := s.char
+		if c == '\\' {
+			s.next()
+			c = s.char
+			if v := decodeEsc(c); v != 0 {
+				c = v
+			} else {
+				switch c {
+				case '\n':
+					nEscape++
+					s.next()
+				case '0', '1', '2', '3', '4', '5', '6', '7':
+					var n int
+					n, c = decodeOctalEsc(s)
+					for i := 1; i < n; i++ {
+						nEscape++
+						s.next()
+					}
+				case 'x':
+					nEscape++
+					s.next()
+					var n int
+					n, c = decodeHexEsc(s)
+					for i := 1; i < n; i++ {
+						nEscape++
+						s.next()
+					}
+				case 'C':
+					nEscape++
+					s.next()
+					if s.char == '-' {
+						nEscape++
+						s.next()
+						c = decodeEsc(s.char)
+					} else {
+						s.failf("invalid escape")
+					}
+				case 'c':
+					nEscape++
+					s.next()
+					c = decodeEsc(s.char)
+				}
+			}
+			nEscape++
+		}
+		s.src[s.offset-nEscape] = c
 		s.next()
 	}
-	s.next()
-	return token.StringPart, s.src[s.begin:s.offset]
+	return s.char, nEscape
+}
+
+func decodeEsc(c byte) byte {
+	return escapes[c]
+}
+
+func decodeCtrlEsc(c byte) byte {
+	if c == '?' {
+		return 0x7f
+	}
+	if v := decodeEsc(c); v != 0 {
+		c = v
+	}
+	return c & 0x9f
+}
+
+func decodeOctalEsc(s *Scanner) (n int, v byte) {
+	for n = 0; n+s.offset < len(s.src) && n < 3; n++ {
+		c := s.src[s.offset+n]
+		if !token.IsOctadecimal(c) {
+			break
+		}
+		v = v*8 + (c - '0')
+	}
+	if n == 0 {
+		s.failf("invalid octal escape")
+	}
+	return
+}
+
+func decodeHexEsc(s *Scanner) (n int, v byte) {
```

Comment:
__[gometalinter]__
* [gocyclo] cyclomatic complexity 11 of function decodeHexEsc() is high (> 10)

<!-- Comment by [SideCI](https://sideci.com) -->